clang/lib/Analysis/UnsafeBufferUsage.cpp

```diff
@@ -1028,6 +1028,50 @@ class UPCPreIncrementGadget : public FixableGadget {
   }
 };
 
+// Representing a pointer type expression of the form `Ptr += n` in an
+// Unspecified Untyped Context (UUC):
+class UUCAddAssignGadget : public FixableGadget {
+private:
+  static constexpr const char *const UUCAddAssignTag =
+      "PointerAddAssignUnderUUC";
+  static constexpr const char *const IntOffsetTag = "IntOffset";
+  static constexpr const char *const OffsetTag = "Offset";
+
+  const BinaryOperator *Node; // the `Ptr += n` node
+  const IntegerLiteral *IntOffset = nullptr;
+  const DeclRefExpr *Offset = nullptr;
+
+public:
+  UUCAddAssignGadget(const MatchFinder::MatchResult &Result)
+      : FixableGadget(Kind::UUCAddAssign),
+        Node(Result.Nodes.getNodeAs<BinaryOperator>(UUCAddAssignTag)),
+        IntOffset(Result.Nodes.getNodeAs<IntegerLiteral>(IntOffsetTag)),
+        Offset(Result.Nodes.getNodeAs<DeclRefExpr>(OffsetTag)) {
+    assert(Node != nullptr && "Expecting a non-null matching result");
+  }
+
+  static bool classof(const Gadget *G) {
+    return G->getKind() == Kind::UUCAddAssign;
+  }
+
+  static Matcher matcher() {
+    return stmt(isInUnspecifiedUntypedContext(expr(ignoringImpCasts(
+        binaryOperator(
+            hasOperatorName("+="), hasLHS(declRefExpr(toSupportedVariable())),
+            hasRHS(expr(anyOf(ignoringImpCasts(declRefExpr().bind(OffsetTag)),
+                              integerLiteral().bind(IntOffsetTag)))))
```

Comment:
Does the exact shape of the offset expression really matter? Why not simply preserve it textually regardless of shape?

Reply:
I make this differentiation to ensure we do not suggest a fixit if the integer literal happens to be negative. Although there's no such check in the case of a `DeclRefExpr`.

Reply:
Oh right! `ULCArraySubscriptGadget` already has a similar facility for determining if the value is non-negative. It's probably good idea to reuse code, because this is the exact same problem. IIRC, @ziqingluo-90 handled the non-constant case by confirming that the expression has unsigned integer type, which sounds like an ok compromise. Also constant-folded the integer before checking.

Might be a good idea to define a matcher "`isNonNegativeIntegerExpr()`" or something, and include it in both gadgets.

I also wonder if implicit casts need to be ignored in both cases. IIRC array subscript expressions don't do implicit promotion, they just stuff arbitrary integer types into the index. But for arithmetic there will be implicit casts to represent promotions, so they may need to be skipped(?)

Reply:
For RHS,  if it is not a constant,  all we need to know about it is whether it is an unsigned integer.    It doesn't have to be a DRE, e.g., `p += (x + 1)` can be transformed to `p = p.subspan(x + 1)` as long as `(x + 1)` has unsigned integer type.